tutorial/lesson_15_generators.cpp

```diff
@@ -0,0 +1,172 @@
+// bash lesson_15_generators_usage.sh
+
+// On os x:
+// g++ lesson_15*.cpp ../tools/GenGen.cpp -g -std=c++11 -fno-rtti -I ../include -L ../bin -lHalide -o lesson_15_generate
+// bash lesson_15_generators_usage.sh
+
+// If you have the entire Halide source tree, you can also build it by
+// running:
+//    make tutorial_lesson_15_generators
+// in a shell with the current directory at the top of the halide
+// source tree.
+
+#include "Halide.h"
+#include <stdio.h>
+
+using namespace Halide;
+
+// Generators are a more structured way to do ahead-of-time
+// compilation of Halide pipelines. Instead of writing an int main()
+// with an ad-hoc command-line interface like we did in lesson 10, we
+// define a class that inherits from Halide::Generator.
+class MyFirstGenerator : public Halide::Generator<MyFirstGenerator> {
+public:
+    // We declare the parameters to the Halide pipeline as public
+    // member variables. We'll give the parameters explicit names this
+    // time. They'll appear in the signature of our generated function
+    // in the same order as we declare them.
+    Param<uint8_t> offset {"offset"};
+    ImageParam input {UInt(8), 2, "input"};
+
+    // Typically you declare your Vars at this scope as well, so that
+    // they can be used in any helper methods you add later.
+    Var x, y;
+
+    // We then define a method that constructs and return the Halide
+    // pipeline:
+    Func build() {
+        // Define the Func.
+        Func brighter;
+        brighter(x, y) = input(x, y) + offset;
+
+        // Schedule it.
+        brighter.vectorize(x, 16).parallel(y);
+
+        // In lesson 10, here is where we called
+        // Func::compile_to_file. In a Generator, we just need to
+        // return the Func representing the output of the pipeline.
+        return brighter;
+    }
+};
+
+// We compile this file along with tools/GenGen.cpp. That file defines
+// an "int main(...)" that provides the command-line interface to use
+// your generator class. We need to tell that code about our
+// generator. We do this like so:
+RegisterGenerator<MyFirstGenerator> my_first_generator{"my_first_generator"};
+
+// If you like, you can put multiple Generators in the one file. This
+// could be a good idea if they share some common code. Let's define
+// another more complex generator:
+class MySecondGenerator : public Halide::Generator<MySecondGenerator> {
+public:
+    // This generator will take some compile-time parameters
+    // too. These let you compile multiple variants of a Halide
+    // pipeline. We'll define one that tells us whether or not to
+    // parallelize in our schedule:
+    GeneratorParam<bool> parallel {"parallel", true /* default value */};
+
+    // ... and another representing a constant scale factor to use:
+    GeneratorParam<float> scale {"scale",
+            1.0f /* default value */,
+            0.0f /* minimum value */,
+            100.0f /* maximum value */};
```

Comment:
It would be nice to show an example of default/min/max values for Params as well.